### ISSUE TITLE
Middleware redirect does not work if the page file doesn't exist

### ISSUE THREAD

**Issue by raulfdm** [withastro/astro#7752]
### What version of `astro` are you using?

2.9.0

### Are you using an SSR adapter? If so, which one?

Not relevant

### What package manager are you using?

pnpm

### What operating system are you using?

Stackblitz

### What browser are you using?

Not relevant

### Describe the Bug

I have a project where I always want to redirect the user from the entry point (`/`) to a locale page (`/en`).

If the `src/pages/index.astro` doesn't exist, this redirect won't happen if we do this via middleware.

Is this correct? I searched in the documentation but I haven't found anything.

https://github.com/withastro/astro/assets/12464600/43b1fc09-2e1a-4037-b4e0-8a209e8e7ef5

### What's the expected result?

No need to have an empty index to I can redirect.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-rfraau?file=src%2Fmiddleware.ts

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by bmenant** [comment]
Similar issue occurs when trying to enforce trailing slash from a middleware, e.g. to redirect `/about` to `/about/` (or vice versa).
If Astro is configured with `trailingSlash: 'always'` then the redirect never happens and a (blank) 404 page is returned.
If Astro is configured with `trailingSlash: 'ignore'` then the redirect works fine.

**Comment by fprl** [comment]
I would like to see middleware running before Astro returning `/404`.

In this case, can't you use [redirects](https://docs.astro.build/en/core-concepts/routing/#configured-redirects)?

**Comment by ematipico** [comment]
This seems to be a regression.

**Comment by syahzuan** [comment]
Can I confirm that this was fixed? I see that this was patched in 2.9.5 and 2.9.6, but I am using v2.10.14 and still have the same issue.
Trying to redirect `/` (which does not exist) to the localized content (`/en` or `/zh` etc), but Astro returns 404 instead of running the middleware.
Trailing slash issue mentioned above by @bmenant also still persists.

Rule-out diagnosis:
1. I tested this with other existing pages and my middleware works fine. 
2. I am using Cloudflare adapter, but issue persists when even when turned off.
3. Tested both "server" or "hybrid", issue persists.
